akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -138,13 +138,21 @@ object Timers {
 
   }
 
-  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration) extends GraphStage[BidiShape[I, I, O, O]] {
+  /**
+   * @param timeoutToTermination allows customising the behaviour once a timeout triggers.
+   *                             Returning None causes completion, and Some(exception) causes failure of the stage with that exception.
+   */
+  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration, timeoutToTermination: () ⇒ Option[Throwable]) extends GraphStage[BidiShape[I, I, O, O]] {
```

Comment:
> The timer stages only log failures themselves if there is a throw, not a failStage().

It is not about the timer stage logging. The reason the change was to it was in order to make the exception more understandable, just using `TimeoutException` for all kinds of timeouts makes it impossible to provide better messages for Akka HTTP since it would have to guess what the reason for the timeouts was. We KNOW it was an idle-timeout and Akka HTTP should log this properly. So it needs information about that.

> Using recover down the line allows to change the error into a normal completion. But again, I don't see why this is needed. 

I don't want completion. I want specific exception types thanks to which I know if it was a tcp idle timeout or some random timeout which I don't care about. This would solve https://github.com/akka/akka-http/issues/651 because I could match on the *specific* exception and give a proper error: https://github.com/akka/akka-http/pull/650/files#diff-623ef36775342f2db6b86855f7ffea64R418 hinting to people that they can change the *specific* setting to tune this behaviour. 

This is a real problem and we must provide better information there. People bump into this a lot, esp with keeping a websocket open, and it closes on them with "no elements passed" they have no idea what's going on, we should provide `oh yeah there was tcp idle timeout on connection to 2.2.2.2 last request was GET /chat, you can tune this with idle-timeout settings`. I need a way of implementing this, without causing Akka Stream TCP to log errors and panic. Because most of the time this timeout is _fine_ and no one should get errors from that.